File: winlux/src/winlux/llm/error_messages_vi.py
"""
Vietnamese Error Messages for Python AI Engines (T61 — REQ-23)

Provides Vietnamese translations for common AI engine error responses.
All Python AI engines (SmartBuy, TrendBrief, CareMate, FIN Tax, DoctorCar, Childhood)
should use this module to return user-friendly Vietnamese messages.

Usage:
  from winlux.llm.error_messages_vi import get_vietnamese_error, vietnamese_error_response

  # In Flask/FastAPI error handlers:
  @app.errorhandler(400)
  def handle_400(e):
      return vietnamese_error_response(str(e), 400)

  # Direct lookup:
  msg = get_vietnamese_error("model not available")
  # Returns: "Hệ thống AI đang bận — vui lòng thử lại sau"
"""

from typing import Optional


# ═══════════════════════════════════════════════════════════════
#  Error Message Map
# ═══════════════════════════════════════════════════════════════

ERROR_MAP: dict[str, dict[str, str | int]] = {
    # ─── Input Validation ───
    "text is required": {"vi": "Vui lòng nhập nội dung", "status": 400},
    "topic is required": {"vi": "Vui lòng nhập chủ đề", "status": 400},
    "query is required": {"vi": "Vui lòng nhập câu hỏi", "status": 400},
    "symptoms is required": {"vi": "Vui lòng mô tả triệu chứng", "status": 400},
    "invalid input": {"vi": "Dữ liệu gửi lên không hợp lệ", "status": 400},
    "text too long": {"vi": "Nội dung quá dài — tối đa 2000 ký tự", "status": 400},
    "text too short": {"vi": "Nội dung quá ngắn — vui lòng mô tả chi tiết hơn", "status": 400},
    "unsupported language": {"vi": "Ngôn ngữ chưa được hỗ trợ", "status": 400},

    # ─── AI / LLM Errors ───
    "model not available": {"vi": "Hệ thống AI đang bận — vui lòng thử lại sau", "status": 503},
    "ollama not available": {"vi": "AI đang khởi động — vui lòng đợi 30 giây rồi thử lại", "status": 503},
    "llm timeout": {"vi": "AI đang xử lý lâu hơn bình thường — vui lòng thử lại", "status": 504},
    "generation failed": {"vi": "Không thể tạo câu trả lời — vui lòng thử lại", "status": 500},
    "context too long": {"vi": "Nội dung quá phức tạp — vui lòng chia nhỏ câu hỏi", "status": 400},
    "rate limited": {"vi": "Quá nhiều yêu cầu — vui lòng đợi 1 phút", "status": 429},
    "circuit breaker open": {"vi": "Hệ thống AI đang bảo trì — sẽ hoạt động lại trong vài phút", "status": 503},

    # ─── RAG / Search ───
    "no results found": {"vi": "Không tìm thấy kết quả phù hợp", "status": 404},
    "embedding failed": {"vi": "Không thể xử lý nội dung — vui lòng thử lại", "status": 500},
    "vector store unavailable": {"vi": "Hệ thống tìm kiếm đang bận — thử lại sau", "status": 503},

    # ─── Product-Specific ───
    # CareMate
    "symptom not recognized": {"vi": "Không nhận diện được triệu chứng — vui lòng mô tả chi tiết hơn", "status": 400},
    "drug not found": {"vi": "Không tìm thấy thuốc trong cơ sở dữ liệu", "status": 404},
    "pharmacy not found": {"vi": "Không tìm thấy nhà thuốc trong khu vực", "status": 404},

    # FIN Tax
    "invalid amount": {"vi": "Số tiền không hợp lệ", "status": 400},
    "tax calculation failed": {"vi": "Không thể tính thuế — vui lòng kiểm tra dữ liệu nhập", "status": 500},
    "ocr failed": {"vi": "Không thể đọc hóa đơn — vui lòng chụp rõ hơn", "status": 400},

    # Doctor Car
    "diagnosis failed": {"vi": "Không thể chẩn đoán — vui lòng mô tả triệu chứng rõ hơn", "status": 500},
    "dtc code not found": {"vi": "Mã lỗi OBD-II không nhận dạng được", "status": 404},
    "vehicle not supported": {"vi": "Hãng xe chưa được hỗ trợ", "status": 400},

    # SmartBuy
    "product not found": {"vi": "Sản phẩm không tồn tại hoặc đã hết hàng", "status": 404},
    "comparison failed": {"vi": "Không thể so sánh — vui lòng thử lại", "status": 500},

    # TrendBrief
    "article not found": {"vi": "Bài viết không tồn tại", "status": 404},
    "summarization failed": {"vi": "Không thể tóm tắt bài viết — vui lòng thử lại", "status": 500},

    # Video Engine
    "script generation failed": {"vi": "Không thể tạo script — vui lòng thử lại", "status": 500},
    "composition failed": {"vi": "Không thể tạo video — vui lòng thử lại", "status": 500},
    "no trending topics": {"vi": "Không tìm thấy chủ đề trending phù hợp", "status": 404},

    # ─── Server / Generic ───
    "internal server error": {"vi": "Hệ thống gặp sự cố — vui lòng thử lại sau", "status": 500},
    "service unavailable": {"vi": "Dịch vụ đang bảo trì — vui lòng quay lại sau", "status": 503},
    "not found": {"vi": "Không tìm thấy nội dung yêu cầu", "status": 404},
    "method not allowed": {"vi": "Phương thức không được hỗ trợ", "status": 405},
}


# ═══════════════════════════════════════════════════════════════
#  Public API
# ═══════════════════════════════════════════════════════════════

def get_vietnamese_error(english_error: str, status_code: Optional[int] = None) -> str:
    """
    Get Vietnamese error message for an English error string.
    Falls back to generic message if no match found.
    """
    error_lower = (english_error or "").lower().strip()

    # Exact match
    if error_lower in ERROR_MAP:
        return ERROR_MAP[error_lower]["vi"]

    # Partial match (contains keyword)
    for key, value in ERROR_MAP.items():
        if key in error_lower:
            return value["vi"]

    # Status-based fallback
    if status_code:
        fallbacks = {
            400: "Dữ liệu gửi lên không hợp lệ",
            401: "Vui lòng đăng nhập",
            403: "Không có quyền truy cập",
            404: "Không tìm thấy",
            429: "Quá nhiều yêu cầu — vui lòng đợi",
            500: "Hệ thống gặp sự cố — vui lòng thử lại sau",
            503: "Dịch vụ đang bận — vui lòng thử lại sau",
        }
        return fallbacks.get(status_code, "Đã xảy ra lỗi — vui lòng thử lại")

    return "Đã xảy ra lỗi — vui lòng thử lại sau"


def vietnamese_error_response(error: str, status_code: int = 500) -> dict:
    """
    Build a standardized error response dict with Vietnamese message.
    Use in Flask/FastAPI error handlers.

    Returns: {"error": "english msg", "message_vi": "Vietnamese msg", "status": code}
    """
    return {
        "error": error,
        "message_vi": get_vietnamese_error(error, status_code),
        "status": status_code,
    }

File: core/src/clients/middleware/error-messages-vi.ts
/**
 * Vietnamese Error Messages (T60 — REQ-23)
 *
 * Shared error message map for all Node.js services.
 * Returns Vietnamese user-friendly messages alongside English technical errors.
 *
 * Usage in Express error handler:
 *   import { getVietnameseError } from '@winlux/service-clients/middleware/error-messages-vi';
 *
 *   app.use((err, req, res, next) => {
 *     const { message_vi, status } = getVietnameseError(err.message, err.status);
 *     res.status(status).json({ error: err.message, message_vi, status });
 *   });
 */

export interface VietnameseError {
  message_vi: string;
  status: number;
}

/**
 * Map English error messages/codes to Vietnamese user-facing messages.
 */
const ERROR_MAP: Record<string, VietnameseError> = {
  // ─── Auth (401/403) ───
  'unauthorized': { message_vi: 'Vui lòng đăng nhập để tiếp tục', status: 401 },
  'token expired': { message_vi: 'Phiên đăng nhập đã hết hạn — vui lòng đăng nhập lại', status: 401 },
  'invalid token': { message_vi: 'Phiên đăng nhập không hợp lệ — vui lòng đăng nhập lại', status: 401 },
  'forbidden': { message_vi: 'Bạn không có quyền thực hiện thao tác này', status: 403 },
  'insufficient permissions': { message_vi: 'Bạn không có quyền truy cập mục này', status: 403 },

  // ─── Validation (400) ───
  'topic is required': { message_vi: 'Vui lòng nhập chủ đề', status: 400 },
  'topics array required': { message_vi: 'Vui lòng chọn ít nhất một chủ đề', status: 400 },
  'email is required': { message_vi: 'Vui lòng nhập email', status: 400 },
  'invalid email': { message_vi: 'Email không hợp lệ', status: 400 },
  'password is required': { message_vi: 'Vui lòng nhập mật khẩu', status: 400 },
  'password too short': { message_vi: 'Mật khẩu phải có ít nhất 6 ký tự', status: 400 },
  'phone is required': { message_vi: 'Vui lòng nhập số điện thoại', status: 400 },
  'invalid phone': { message_vi: 'Số điện thoại không hợp lệ', status: 400 },
  'otp is required': { message_vi: 'Vui lòng nhập mã OTP', status: 400 },
  'invalid otp': { message_vi: 'Mã OTP không đúng hoặc đã hết hạn', status: 400 },
  'missing required fields': { message_vi: 'Vui lòng điền đầy đủ thông tin bắt buộc', status: 400 },
  'invalid amount': { message_vi: 'Số tiền không hợp lệ', status: 400 },
  'invalid date': { message_vi: 'Ngày không hợp lệ', status: 400 },

  // ─── Not Found (404) ───
  'not found': { message_vi: 'Không tìm thấy nội dung yêu cầu', status: 404 },
  'video not found': { message_vi: 'Video không tồn tại', status: 404 },
  'product not found': { message_vi: 'Sản phẩm không tồn tại', status: 404 },
  'article not found': { message_vi: 'Bài viết không tồn tại', status: 404 },
  'user not found': { message_vi: 'Tài khoản không tồn tại', status: 404 },
  'channel not found': { message_vi: 'Kênh không tồn tại', status: 404 },
  'pharmacy not found': { message_vi: 'Nhà thuốc không tồn tại', status: 404 },
  'drug not found': { message_vi: 'Thuốc không tìm thấy', status: 404 },
  'garage not found': { message_vi: 'Garage không tồn tại', status: 404 },
  'transaction not found': { message_vi: 'Giao dịch không tồn tại', status: 404 },
  'vehicle not found': { message_vi: 'Xe không tồn tại trong hệ thống', status: 404 },

  // ─── Conflict (409) ───
  'email already exists': { message_vi: 'Email đã được sử dụng — thử đăng nhập hoặc dùng email khác', status: 409 },
  'phone already exists': { message_vi: 'Số điện thoại đã được đăng ký', status: 409 },
  'already subscribed': { message_vi: 'Bạn đã đăng ký gói này rồi', status: 409 },
  'duplicate': { message_vi: 'Dữ liệu đã tồn tại', status: 409 },

  // ─── Rate Limit (429) ───
  'rate limit exceeded': { message_vi: 'Bạn đã gửi quá nhiều yêu cầu — vui lòng thử lại sau 1 phút', status: 429 },
  'too many requests': { message_vi: 'Quá nhiều yêu cầu — vui lòng đợi một lát', status: 429 },
  'daily limit reached': { message_vi: 'Đã hết lượt sử dụng miễn phí hôm nay — nâng cấp Pro để dùng không giới hạn', status: 429 },

  // ─── Payment (402) ───
  'payment required': { message_vi: 'Tính năng này yêu cầu gói Pro — nâng cấp ngay', status: 402 },
  'payment failed': { message_vi: 'Thanh toán thất bại — vui lòng thử lại hoặc đổi phương thức', status: 402 },
  'subscription expired': { message_vi: 'Gói đăng ký đã hết hạn — gia hạn để tiếp tục', status: 402 },

  // ─── AI / Processing (500/503) ───
  'ai service unavailable': { message_vi: 'Hệ thống AI đang bận — vui lòng thử lại trong giây lát', status: 503 },
  'ai timeout': { message_vi: 'AI đang xử lý lâu hơn bình thường — vui lòng thử lại', status: 504 },
  'generation failed': { message_vi: 'Không thể tạo nội dung — vui lòng thử lại', status: 500 },
  'processing error': { message_vi: 'Đã xảy ra lỗi khi xử lý — vui lòng thử lại sau', status: 500 },

  // ─── Server (500) ───
  'internal server error': { message_vi: 'Hệ thống gặp sự cố — vui lòng thử lại sau', status: 500 },
  'database error': { message_vi: 'Lỗi kết nối dữ liệu — vui lòng thử lại', status: 500 },
  'service unavailable': { message_vi: 'Dịch vụ đang bảo trì — vui lòng quay lại sau', status: 503 },
};

/**
 * Get Vietnamese error message for a given English error.
 * Falls back to generic Vietnamese message if no exact match found.
 */
export function getVietnameseError(englishError: string, statusCode?: number): VietnameseError {
  const errorLower = (englishError || '').toLowerCase().trim();

  // Exact match
  if (ERROR_MAP[errorLower]) {
    return ERROR_MAP[errorLower];
  }

  // Partial match (contains key phrase)
  for (const [key, value] of Object.entries(ERROR_MAP)) {
    if (errorLower.includes(key)) {
      return value;
    }
  }

  // Status code based fallback
  const status = statusCode || 500;
  if (status === 400) return { message_vi: 'Dữ liệu gửi lên không hợp lệ', status: 400 };
  if (status === 401) return { message_vi: 'Vui lòng đăng nhập', status: 401 };
  if (status === 403) return { message_vi: 'Không có quyền truy cập', status: 403 };
  if (status === 404) return { message_vi: 'Không tìm thấy', status: 404 };
  if (status === 429) return { message_vi: 'Quá nhiều yêu cầu — vui lòng đợi', status: 429 };

  // Generic fallback
  return { message_vi: 'Đã xảy ra lỗi — vui lòng thử lại sau', status };
}

/**
 * Express error handler middleware with Vietnamese messages.
 * Drop this into any product service's middleware stack:
 *
 *   import { vietnameseErrorHandler } from '@winlux/service-clients/middleware/error-messages-vi';
 *   app.use(vietnameseErrorHandler);
 */
export function vietnameseErrorHandler(err: any, req: any, res: any, next: any): void {
  const status = err.status || err.statusCode || 500;
  const message = err.message || 'Internal Server Error';
  const { message_vi } = getVietnameseError(message, status);

  // Log for debugging (English)
  if (status >= 500) {
    console.error(`[${req.method} ${req.path}] ${status}: ${message}`, err.stack || '');
  }

  res.status(status).json({
    error: message,
    message_vi,
    status,
    ...(process.env.NODE_ENV === 'development' ? { stack: err.stack } : {}),
  });
}
